File: packages/dvai-bridge-capacitor-llama/ios/Tests/DVAICapacitorLlamaTests/LlamaHandlersTest.swift
import XCTest
@testable import DVAICapacitorLlama
import DVAICapacitorLlamaObjC

/// Mock bridge so handler tests don't need a real GGUF model loaded.
/// Records the prompt that was passed in and returns canned values.
final class MockBridge: LlamaCppBridgeProtocol {
    var loaded: Bool = true
    var completionToReturn: String = "canned response"
    var embeddingToReturn: [NSNumber] = [NSNumber(value: 0.1), NSNumber(value: 0.2), NSNumber(value: 0.3)]
    var receivedPrompt: String?
    var receivedEmbeddingTexts: [String] = []
    var completionShouldThrow: Bool = false
    var embeddingShouldThrow: Bool = false

    var isLoaded: Bool { loaded }

    func completePrompt(_ prompt: String, maxTokens: Int32, temperature: Float, topP: Float) throws -> String {
        receivedPrompt = prompt
        if completionShouldThrow {
            throw NSError(domain: "MockBridge", code: 99, userInfo: [NSLocalizedDescriptionKey: "boom"])
        }
        return completionToReturn
    }

    func embedding(_ text: String) throws -> [NSNumber] {
        receivedEmbeddingTexts.append(text)
        if embeddingShouldThrow {
            throw NSError(domain: "MockBridge", code: 100, userInfo: [NSLocalizedDescriptionKey: "embed boom"])
        }
        return embeddingToReturn
    }
}

final class LlamaHandlersTest: XCTestCase {
    let ctx = HandlerContext(modelId: "test-model", backendName: "llama")

    private func makeHandlers(
        bridge: MockBridge = MockBridge(),
        mmprojLoaded: Bool = false,
        modelHasAudioEncoder: Bool = false,
        embeddingMode: Bool = false
    ) -> LlamaHandlers {
        LlamaHandlers(
            bridgeProtocol: bridge,
            modelId: "test-model",
            mmprojLoaded: mmprojLoaded,
            modelHasAudioEncoder: modelHasAudioEncoder,
            embeddingMode: embeddingMode
        )
    }

    // MARK: - Chat completion

    func testChatCompletionTextHappyPath() async throws {
        let bridge = MockBridge()
        bridge.completionToReturn = "Hello, world!"
        let handlers = makeHandlers(bridge: bridge)
        let body: [String: Any] = ["messages": [["role": "user", "content": "hi"]]]
        let resp = try await handlers.handleChatCompletion(body: body, ctx: ctx)
        guard case .json(let status, let bodyAny) = resp else {
            XCTFail("expected .json response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 200)
        let json = bodyAny as? [String: Any]
        XCTAssertEqual(json?["object"] as? String, "chat.completion")
        XCTAssertEqual(json?["model"] as? String, "test-model")
        let choices = json?["choices"] as? [[String: Any]]
        let msg = choices?.first?["message"] as? [String: Any]
        XCTAssertEqual(msg?["content"] as? String, "Hello, world!")
        XCTAssertEqual(msg?["role"] as? String, "assistant")
        XCTAssertEqual(choices?.first?["finish_reason"] as? String, "stop")
        XCTAssertEqual(bridge.receivedPrompt, "hi")
    }

    func testChatCompletionStreamingTextEmitsRoleContentFinishDone() async throws {
        let bridge = MockBridge()
        bridge.completionToReturn = "stream-canned"
        let handlers = makeHandlers(bridge: bridge)
        let body: [String: Any] = [
            "messages": [["role": "user", "content": "hi"]],
            "stream": true,
        ]
        let resp = try await handlers.handleChatCompletion(body: body, ctx: ctx)
        guard case .sse(let stream) = resp else {
            XCTFail("expected .sse response, got \(resp)")
            return
        }
        var collected: [String] = []
        for await chunk in stream {
            collected.append(chunk)
        }
        XCTAssertEqual(collected.count, 4, "expected 4 SSE frames: role / content / finish / [DONE]")
        // Frame 0: role delta
        XCTAssertTrue(collected[0].contains("\"role\":\"assistant\""), "frame 0 should be role delta: \(collected[0])")
        // Frame 1: content delta with the canned string
        XCTAssertTrue(collected[1].contains("stream-canned"), "frame 1 should contain content: \(collected[1])")
        // Frame 2: finish chunk
        XCTAssertTrue(collected[2].contains("\"finish_reason\":\"stop\""), "frame 2 should have finish_reason: \(collected[2])")
        // Frame 3: [DONE]
        XCTAssertEqual(collected[3], "data: [DONE]\n\n")
    }

    func testChatCompletionImageWithoutMmprojReturns400() async throws {
        let handlers = makeHandlers(mmprojLoaded: false)
        let body: [String: Any] = [
            "messages": [[
                "role": "user",
                "content": [
                    ["type": "image_url", "image_url": ["url": "data:image/png;base64,iVBOR"]] as [String: Any],
                ],
            ] as [String: Any]],
        ]
        let resp = try await handlers.handleChatCompletion(body: body, ctx: ctx)
        guard case .error(let status, let message) = resp else {
            XCTFail("expected .error response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 400)
        XCTAssertTrue(message.contains("no mmproj was loaded"), "message: \(message)")
        XCTAssertTrue(message.contains("nativeMmprojPath"), "message: \(message)")
    }

    func testChatCompletionAudioWithoutEncoderReturns400() async throws {
        let handlers = makeHandlers(modelHasAudioEncoder: false)
        let body: [String: Any] = [
            "messages": [[
                "role": "user",
                "content": [
                    ["type": "input_audio", "input_audio": ["data": "AAAA", "format": "pcm16"]] as [String: Any],
                ],
            ] as [String: Any]],
        ]
        let resp = try await handlers.handleChatCompletion(body: body, ctx: ctx)
        guard case .error(let status, let message) = resp else {
            XCTFail("expected .error response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 400)
        XCTAssertTrue(message.contains("native audio encoder"), "message: \(message)")
    }

    func testChatCompletionMissingMessagesReturns400() async throws {
        let handlers = makeHandlers()
        let resp = try await handlers.handleChatCompletion(body: [:], ctx: ctx)
        guard case .error(let status, let message) = resp else {
            XCTFail("expected .error response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 400)
        XCTAssertTrue(message.contains("messages"), "message: \(message)")
    }

    // MARK: - Legacy /v1/completions

    func testCompletionLegacyConvertsToTextCompletion() async throws {
        let bridge = MockBridge()
        bridge.completionToReturn = "canned-text"
        let handlers = makeHandlers(bridge: bridge)
        let resp = try await handlers.handleCompletion(body: ["prompt": "say hi"], ctx: ctx)
        guard case .json(let status, let bodyAny) = resp else {
            XCTFail("expected .json response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 200)
        let json = bodyAny as? [String: Any]
        XCTAssertEqual(json?["object"] as? String, "text_completion")
        let choices = json?["choices"] as? [[String: Any]]
        XCTAssertEqual(choices?.first?["text"] as? String, "canned-text")
        XCTAssertEqual(choices?.first?["finish_reason"] as? String, "stop")
        // logprobs is NSNull — present in the dict.
        XCTAssertNotNil(choices?.first?["logprobs"])
        // ID was rewritten chatcmpl- → cmpl-
        let idStr = json?["id"] as? String ?? ""
        XCTAssertTrue(idStr.hasPrefix("cmpl-"), "id should start with cmpl-: \(idStr)")
        // Round-tripped prompt
        XCTAssertEqual(bridge.receivedPrompt, "say hi")
    }

    func testCompletionLegacyArrayPromptJoinedWithNewline() async throws {
        let bridge = MockBridge()
        let handlers = makeHandlers(bridge: bridge)
        let resp = try await handlers.handleCompletion(body: ["prompt": ["alpha", "beta"]], ctx: ctx)
        guard case .json = resp else {
            XCTFail("expected .json response, got \(resp)")
            return
        }
        XCTAssertEqual(bridge.receivedPrompt, "alpha\nbeta")
    }

    // MARK: - Embeddings

    func testEmbeddingsRejectedWhenNotEmbeddingMode() async throws {
        let handlers = makeHandlers(embeddingMode: false)
        let resp = try await handlers.handleEmbeddings(body: ["input": "hello"], ctx: ctx)
        guard case .error(let status, let message) = resp else {
            XCTFail("expected .error response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 400)
        XCTAssertTrue(message.contains("nativeEmbeddingMode"), "message: \(message)")
    }

    func testEmbeddingsHappyPathSingleString() async throws {
        let bridge = MockBridge()
        bridge.embeddingToReturn = [NSNumber(value: 0.5), NSNumber(value: -0.25), NSNumber(value: 1.0)]
        let handlers = makeHandlers(bridge: bridge, embeddingMode: true)
        let resp = try await handlers.handleEmbeddings(body: ["input": "hello"], ctx: ctx)
        guard case .json(let status, let bodyAny) = resp else {
            XCTFail("expected .json response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 200)
        let json = bodyAny as? [String: Any]
        XCTAssertEqual(json?["object"] as? String, "list")
        XCTAssertEqual(json?["model"] as? String, "test-model")
        let data = json?["data"] as? [[String: Any]]
        XCTAssertEqual(data?.count, 1)
        XCTAssertEqual(data?.first?["object"] as? String, "embedding")
        XCTAssertEqual(data?.first?["index"] as? Int, 0)
        let vec = data?.first?["embedding"] as? [Double]
        XCTAssertEqual(vec?.count, 3)
        XCTAssertEqual(vec?[0], 0.5, accuracy: 1e-9)
        XCTAssertEqual(vec?[1], -0.25, accuracy: 1e-9)
        XCTAssertEqual(vec?[2], 1.0, accuracy: 1e-9)
        XCTAssertEqual(bridge.receivedEmbeddingTexts, ["hello"])
    }

    func testEmbeddingsArrayInputProducesMultipleEntries() async throws {
        let bridge = MockBridge()
        let handlers = makeHandlers(bridge: bridge, embeddingMode: true)
        let resp = try await handlers.handleEmbeddings(
            body: ["input": ["alpha", "beta", "gamma"]],
            ctx: ctx
        )
        guard case .json(let status, let bodyAny) = resp else {
            XCTFail("expected .json response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 200)
        let data = (bodyAny as? [String: Any])?["data"] as? [[String: Any]]
        XCTAssertEqual(data?.count, 3)
        XCTAssertEqual(data?[0]["index"] as? Int, 0)
        XCTAssertEqual(data?[1]["index"] as? Int, 1)
        XCTAssertEqual(data?[2]["index"] as? Int, 2)
        XCTAssertEqual(bridge.receivedEmbeddingTexts, ["alpha", "beta", "gamma"])
    }

    func testEmbeddingsMissingInputReturns400() async throws {
        let handlers = makeHandlers(embeddingMode: true)
        let resp = try await handlers.handleEmbeddings(body: [:], ctx: ctx)
        guard case .error(let status, _) = resp else {
            XCTFail("expected .error response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 400)
    }

    // MARK: - Models

    func testModelsReturnsSingleEntryListWithCtxModelId() async throws {
        let handlers = makeHandlers()
        let customCtx = HandlerContext(modelId: "/path/to/model.gguf", backendName: "llama")
        let resp = try await handlers.handleModels(ctx: customCtx)
        guard case .json(let status, let bodyAny) = resp else {
            XCTFail("expected .json response, got \(resp)")
            return
        }
        XCTAssertEqual(status, 200)
        let json = bodyAny as? [String: Any]
        XCTAssertEqual(json?["object"] as? String, "list")
        let data = json?["data"] as? [[String: Any]]
        XCTAssertEqual(data?.count, 1)
        XCTAssertEqual(data?.first?["id"] as? String, "/path/to/model.gguf")
    }
}

File: packages/dvai-bridge-android-llama-core/android/src/test/java/co/deepvoiceai/bridge/llama/core/LlamaHandlersTest.kt
package co.deepvoiceai.bridge.llama.core

import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.add
import kotlinx.serialization.json.addJsonObject
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

/**
 * JVM unit tests for [LlamaHandlers]. Uses a [FakeBridge] implementing
 * [LlamaCppBridgeApi] so tests don't need a real GGUF model loaded.
 */
class LlamaHandlersTest {
    private val ctx = HandlerContext(modelId = "test-model", backendName = "llama")

    private class FakeBridge(
        var loaded: Boolean = true,
        var completionToReturn: String? = "canned response",
        var multimodalCompletionToReturn: String? = "canned multimodal response",
        var embeddingToReturn: FloatArray? = floatArrayOf(0.1f, 0.2f, 0.3f),
    ) : LlamaCppBridgeApi {
        var receivedPrompt: String? = null
        var receivedMultimodalPrompt: String? = null
        var receivedMedia: List<ByteArray> = emptyList()
        val receivedEmbeddingTexts = mutableListOf<String>()
        var mmprojLoadedFlag: Boolean = false
        var receivedMmprojPath: String? = null
        var loadMmprojResult: Boolean = true
        var modelHasAudioEncoder: Boolean = false
        var receivedChatTemplate: String? = null
        var receivedChatMessages: List<Map<String, String>> = emptyList()
        var chatTemplateRenderer: (List<Map<String, String>>, Boolean) -> String? = { msgs, addAssistant ->
            buildString {
                for (m in msgs) {
                    append(m["role"] ?: "user")
                    append(": ")
                    append(m["content"] ?: "")
                    append('\n')
                }
                if (addAssistant) append("assistant:")
            }
        }

        override fun isLoaded(): Boolean = loaded
        override fun completePrompt(prompt: String, maxTokens: Int, temperature: Float, topP: Float): String? {
            receivedPrompt = prompt
            return completionToReturn
        }
        override fun embedding(text: String): FloatArray? {
            receivedEmbeddingTexts += text
            return embeddingToReturn
        }
        override fun loadMmproj(mmprojPath: String): Boolean {
            receivedMmprojPath = mmprojPath
            if (loadMmprojResult) mmprojLoadedFlag = true
            return loadMmprojResult
        }
        override fun unloadMmproj() {
            mmprojLoadedFlag = false
        }
        override fun isMmprojLoaded(): Boolean = mmprojLoadedFlag
        override fun hasAudioEncoder(): Boolean = modelHasAudioEncoder
        override fun applyChatTemplate(
            templateOverride: String?,
            messages: List<Map<String, String>>,
            addAssistant: Boolean,
        ): String? {
            receivedChatTemplate = templateOverride
            receivedChatMessages = messages
            return chatTemplateRenderer(messages, addAssistant)
        }
        override fun completeMultimodalPrompt(
            prompt: String,
            media: List<ByteArray>,
            maxTokens: Int,
            temperature: Float,
            topP: Float,
        ): String? {
            receivedMultimodalPrompt = prompt
            receivedMedia = media
            return multimodalCompletionToReturn
        }
    }

    private fun makeHandlers(
        bridge: FakeBridge = FakeBridge(),
        mmprojLoaded: Boolean = false,
        modelHasAudioEncoder: Boolean = false,
        embeddingMode: Boolean = false,
        chatTemplate: String? = null,
    ): LlamaHandlers = LlamaHandlers(
        bridge = bridge,
        modelId = "test-model",
        mmprojLoaded = mmprojLoaded,
        modelHasAudioEncoder = modelHasAudioEncoder,
        embeddingMode = embeddingMode,
        chatTemplate = chatTemplate,
    )

    // ----- Chat completion -----

    @Test
    fun `chat completion text happy path`() = runBlocking {
        val bridge = FakeBridge(completionToReturn = "Hello, world!")
        val handlers = makeHandlers(bridge = bridge)
        val body = buildJsonObject {
            putJsonArray("messages") {
                addJsonObject {
                    put("role", "user")
                    put("content", "hi")
                }
            }
        }
        val resp = handlers.handleChatCompletion(body, ctx) as? HandlerResponse.Json
            ?: error("expected Json response")
        assertEquals(200, resp.status)
        val obj = resp.body as JsonObject
        assertEquals("chat.completion", (obj["object"] as JsonPrimitive).content)
        assertEquals("test-model", (obj["model"] as JsonPrimitive).content)
        val choices = obj["choices"] as JsonArray
        assertEquals(1, choices.size)
        val msg = (choices[0] as JsonObject)["message"] as JsonObject
        assertEquals("Hello, world!", (msg["content"] as JsonPrimitive).content)
        assertEquals("assistant", (msg["role"] as JsonPrimitive).content)
        assertEquals("stop", ((choices[0] as JsonObject)["finish_reason"] as JsonPrimitive).content)
        // The fake renders `<role>: <content>\n[assistant:]` — we just assert
        // the rendered prompt contains the original user content.
        assertTrue("receivedPrompt should contain 'hi': ${bridge.receivedPrompt}", bridge.receivedPrompt!!.contains("hi"))
    }

    /**
     * Parse a single SSE frame's JSON payload. Returns null for `[DONE]`
     * frames or non-`data:` lines.
     */
    private fun decodeFrame(frame: String): JsonObject? {
        val trimmed = frame.trim()
        if (!trimmed.startsWith("data: ")) return null
        val payload = trimmed.removePrefix("data: ")
        if (payload == "[DONE]") return null
        return Json.parseToJsonElement(payload).jsonObject
    }

    @Test
    fun `chat completion streaming text emits role content finish done`() = runBlocking {
        val bridge = FakeBridge(completionToReturn = "stream-canned")
        val handlers = makeHandlers(bridge = bridge)
        val body = buildJsonObject {
            putJsonArray("messages") {
                addJsonObject {
                    put("role", "user")
                    put("content", "hi")
                }
            }
            put("stream", true)
        }
        val resp = handlers.handleChatCompletion(body, ctx) as? HandlerResponse.Sse
            ?: error("expected Sse response")
        val frames = resp.flow.toList()
        assertEquals(4, frames.size)

        // Frame 0: role delta
        val roleDelta = decodeFrame(frames[0])
            ?.get("choices")?.jsonArray?.first()?.jsonObject
            ?.get("delta")?.jsonObject
        assertEquals("assistant", roleDelta?.get("role")?.jsonPrimitive?.content)

        // Frame 1: content delta
        val contentDelta = decodeFrame(frames[1])
            ?.get("choices")?.jsonArray?.first()?.jsonObject
            ?.get("delta")?.jsonObject
        assertEquals("stream-canned", contentDelta?.get("content")?.jsonPrimitive?.content)

        // Frame 2: finish chunk
        val finishChoice = decodeFrame(frames[2])
            ?.get("choices")?.jsonArray?.first()?.jsonObject
        assertEquals("stop", finishChoice?.get("finish_reason")?.jsonPrimitive?.content)

        // Frame 3: [DONE]
        assertEquals("data: [DONE]\n\n", frames[3])
        assertNull(decodeFrame(frames[3]))
    }

    @Test
    fun `chat completion returns 500 when bridge returns null`() = runBlocking {
        val bridge = FakeBridge(completionToReturn = null)
        val handlers = makeHandlers(bridge = bridge)
        val body = buildJsonObject {
            putJsonArray("messages") {
                addJsonObject {
                    put("role", "user")
                    put("content", "hi")
                }
            }
        }
        val resp = handlers.handleChatCompletion(body, ctx) as? HandlerResponse.Error
            ?: error("expected Error response")
        assertEquals(500, resp.status)
    }

    @Test
    fun `chat completion image without mmproj returns 400`() = runBlocking {
        val handlers = makeHandlers(mmprojLoaded = false)
        val body = buildJsonObject {
            putJsonArray("messages") {
                addJsonObject {
                    put("role", "user")
                    putJsonArray("content") {
                        addJsonObject {
                            put("type", "image_url")
                            putJsonObject("image_url") {
                                put("url", "data:image/png;base64,iVBOR")
                            }
                        }
                    }
                }
            }
        }
        val resp = handlers.handleChatCompletion(body, ctx) as? HandlerResponse.Error
            ?: error("expected Error response")
        assertEquals(400, resp.status)
        assertTrue("message: ${resp.message}", resp.message.contains("no mmproj was loaded"))
        assertTrue("message: ${resp.message}", resp.message.contains("nativeMmprojPath"))
    }

    @Test
    fun `chat completion audio without encoder returns 400`() = runBlocking {
        val handlers = makeHandlers(modelHasAudioEncoder = false)
        val body = buildJsonObject {
            putJsonArray("messages") {
                addJsonObject {
                    put("role", "user")
                    putJsonArray("content") {
                        addJsonObject {
                            put("type", "input_audio")
                            putJsonObject("input_audio") {
                                put("data", "AAAA")
                                put("format", "pcm16")
                            }
                        }
                    }
                }
            }
        }
        val resp = handlers.handleChatCompletion(body, ctx) as? HandlerResponse.Error
            ?: error("expected Error response")
        assertEquals(400, resp.status)
        assertTrue("message: ${resp.message}", resp.message.contains("native audio encoder"))
    }

    // ----- Multimodal happy paths (Phase 2A Pass 2) -----

    @Test
    fun `chat vision happy path`() = runBlocking {
        val bridge = FakeBridge(multimodalCompletionToReturn = "I see a 1x1 transparent pixel.")
        val handlers = makeHandlers(bridge = bridge, mmprojLoaded = true)
        val body = buildJsonObject {
            putJsonArray("messages") {
                addJsonObject {
                    put("role", "user")
                    putJsonArray("content") {
                        addJsonObject {
                            put("type", "text"); put("text", "What's in this image?")
                        }
                        addJsonObject {
                            put("type", "image_url")
                            putJsonObject("image_url") {
                                put("url", "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNk+M9QDwADhgGAWjR9awAAAABJRU5ErkJggg==")
                            }
                        }
                    }
                }
            }
        }
        val resp = handlers.handleChatCompletion(body, ctx) as? HandlerResponse.Json
            ?: error("expected Json response")
        assertEquals(200, resp.status)
        assertEquals(1, bridge.receivedMedia.size)
        val prompt = bridge.receivedMultimodalPrompt!!
        val markerCount = prompt.split(MTMD_MEDIA_MARKER).size - 1
        assertEquals("expected one <__media__> marker, got: $prompt", 1, markerCount)
    }

    @Test
    fun `chat audio happy path`() = runBlocking {
        val bridge = FakeBridge(multimodalCompletionToReturn = "hello world")
        val handlers = makeHandlers(bridge = bridge, modelHasAudioEncoder = true)
        val pcm = ByteArray(32) // 32 zero bytes — pcm16 is pass-through, no decode call
        val pcmB64 = java.util.Base64.getEncoder().encodeToString(pcm)
        val body = buildJsonObject {
            putJsonArray("messages") {
                addJsonObject {
                    put("role", "user")
                    putJsonArray("content") {
                        addJsonObject { put("type", "text"); put("text", "Transcribe:") }
                        addJsonObject {
                            put("type", "input_audio")
                            putJsonObject("input_audio") {
                                put("data", pcmB64)
                                put("format", "pcm16")
                            }
                        }
                    }
                }
            }
        }
        val resp = handlers.handleChatCompletion(body, ctx) as? HandlerResponse.Json
            ?: error("expected Json response")
        assertEquals(200, resp.status)
        assertEquals(1, bridge.receivedMedia.size)
        org.junit.Assert.assertArrayEquals(pcm, bridge.receivedMedia[0])
        val prompt = bridge.receivedMultimodalPrompt!!
        val markerCount = prompt.split(MTMD_MEDIA_MARKER).size - 1
        assertEquals(1, markerCount)
    }

    @Test
    fun `chat interleaved image audio preserves order and marker count`() = runBlocking {
        val bridge = FakeBridge()
        val handlers = makeHandlers(
            bridge = bridge,
            mmprojLoaded = true,
            modelHasAudioEncoder = true,
        )
        val pcm = ByteArray(16) { 0xAB.toByte() }
        val pcmB64 = java.util.Base64.getEncoder().encodeToString(pcm)
        val body = buildJsonObject {
            putJsonArray("messages") {
                addJsonObject {
                    put("role", "user")
                    putJsonArray("content") {
                        addJsonObject { put("type", "text"); put("text", "alpha") }
                        addJsonObject {
                            put("type", "image_url")
                            putJsonObject("image_url") { put("url", "data:image/png;base64,iVBORw0KGgo=") }
                        }
                        addJsonObject { put("type", "text"); put("text", "beta") }
                        addJsonObject {
                            put("type", "input_audio")
                            putJsonObject("input_audio") {
                                put("data", pcmB64)
                                put("format", "pcm16")
                            }
                        }
                        addJsonObject { put("type", "text"); put("text", "gamma") }
                    }
                }
            }
        }
        val resp = handlers.handleChatCompletion(body, ctx) as? HandlerResponse.Json
            ?: error("expected Json response")
        assertEquals(200, resp.status)
        assertEquals(2, bridge.receivedMedia.size)
        // Audio is at index 1 (image first, then audio in declaration order).
        org.junit.Assert.assertArrayEquals(pcm, bridge.receivedMedia[1])
        val prompt = bridge.receivedMultimodalPrompt!!
        val markerCount = prompt.split(MTMD_MEDIA_MARKER).size - 1
        assertEquals("expected two <__media__> markers, got: $prompt", 2, markerCount)
    }

    @Test
    fun `chat completion missing messages returns 400`() = runBlocking {
        val handlers = makeHandlers()
        val resp = handlers.handleChatCompletion(buildJsonObject {}, ctx) as? HandlerResponse.Error
            ?: error("expected Error response")
        assertEquals(400, resp.status)
        assertTrue("message: ${resp.message}", resp.message.contains("messages"))
    }

    // ----- Legacy completions -----

    @Test
    fun `legacy completion converts to text_completion shape`() = runBlocking {
        val bridge = FakeBridge(completionToReturn = "canned-text")
        val handlers = makeHandlers(bridge = bridge)
        val body = buildJsonObject { put("prompt", "say hi") }
        val resp = handlers.handleCompletion(body, ctx) as? HandlerResponse.Json
            ?: error("expected Json response")
        assertEquals(200, resp.status)
        val obj = resp.body as JsonObject
        assertEquals("text_completion", (obj["object"] as JsonPrimitive).content)
        val choices = obj["choices"] as JsonArray
        val choice0 = choices[0] as JsonObject
        assertEquals("canned-text", (choice0["text"] as JsonPrimitive).content)
        assertEquals("stop", (choice0["finish_reason"] as JsonPrimitive).content)
        // logprobs key present (JsonNull)
        assertNotNull(choice0["logprobs"])
        // ID rewritten chatcmpl- → cmpl-
        val idStr = (obj["id"] as JsonPrimitive).content
        assertTrue("id should start with cmpl-: $idStr", idStr.startsWith("cmpl-"))
        assertTrue("receivedPrompt should contain 'say hi': ${bridge.receivedPrompt}", bridge.receivedPrompt!!.contains("say hi"))
    }

    @Test
    fun `legacy completion array prompt joined with newline`() = runBlocking {
        val bridge = FakeBridge()
        val handlers = makeHandlers(bridge = bridge)
        val body = buildJsonObject {
            putJsonArray("prompt") {
                add("alpha")
                add("beta")
            }
        }
        val resp = handlers.handleCompletion(body, ctx) as? HandlerResponse.Json
            ?: error("expected Json response")
        assertEquals(200, resp.status)
        // Prompt is joined alpha\nbeta and then chat-templated by the fake renderer.
        assertTrue("receivedPrompt should contain 'alpha\\nbeta': ${bridge.receivedPrompt}", bridge.receivedPrompt!!.contains("alpha\nbeta"))
    }

    // ----- Embeddings -----

    @Test
    fun `embeddings rejected when not embedding mode`() = runBlocking {
        val handlers = makeHandlers(embeddingMode = false)
        val resp = handlers.handleEmbeddings(
            buildJsonObject { put("input", "hello") },
            ctx,
        ) as? HandlerResponse.Error ?: error("expected Error response")
        assertEquals(400, resp.status)
        assertTrue("message: ${resp.message}", resp.message.contains("nativeEmbeddingMode"))
    }

    @Test
    fun `embeddings happy path single string`() = runBlocking {
        val bridge = FakeBridge(embeddingToReturn = floatArrayOf(0.5f, -0.25f, 1.0f))
        val handlers = makeHandlers(bridge = bridge, embeddingMode = true)
        val resp = handlers.handleEmbeddings(
            buildJsonObject { put("input", "hello") },
            ctx,
        ) as? HandlerResponse.Json ?: error("expected Json response")
        assertEquals(200, resp.status)
        val obj = resp.body as JsonObject
        assertEquals("list", (obj["object"] as JsonPrimitive).content)
        assertEquals("test-model", (obj["model"] as JsonPrimitive).content)
        val data = obj["data"] as JsonArray
        assertEquals(1, data.size)
        val entry0 = data[0] as JsonObject
        assertEquals("embedding", (entry0["object"] as JsonPrimitive).content)
        assertEquals(0, (entry0["index"] as JsonPrimitive).intOrNull)
        val vec = entry0["embedding"] as JsonArray
        assertEquals(3, vec.size)
        assertEquals(0.5, (vec[0] as JsonPrimitive).content.toDouble(), 1e-6)
        assertEquals(-0.25, (vec[1] as JsonPrimitive).content.toDouble(), 1e-6)
        assertEquals(1.0, (vec[2] as JsonPrimitive).content.toDouble(), 1e-6)
        assertEquals(listOf("hello"), bridge.receivedEmbeddingTexts)
    }

    @Test
    fun `embeddings array input produces multiple entries`() = runBlocking {
        val bridge = FakeBridge()
        val handlers = makeHandlers(bridge = bridge, embeddingMode = true)
        val resp = handlers.handleEmbeddings(
            buildJsonObject {
                putJsonArray("input") {
                    add("alpha")
                    add("beta")
                    add("gamma")
                }
            },
            ctx,
        ) as? HandlerResponse.Json ?: error("expected Json response")
        assertEquals(200, resp.status)
        val data = (resp.body as JsonObject)["data"] as JsonArray
        assertEquals(3, data.size)
        assertEquals(0, ((data[0] as JsonObject)["index"] as JsonPrimitive).intOrNull)
        assertEquals(1, ((data[1] as JsonObject)["index"] as JsonPrimitive).intOrNull)
        assertEquals(2, ((data[2] as JsonObject)["index"] as JsonPrimitive).intOrNull)
        assertEquals(listOf("alpha", "beta", "gamma"), bridge.receivedEmbeddingTexts)
    }

    @Test
    fun `embeddings missing input returns 400`() = runBlocking {
        val handlers = makeHandlers(embeddingMode = true)
        val resp = handlers.handleEmbeddings(buildJsonObject {}, ctx) as? HandlerResponse.Error
            ?: error("expected Error response")
        assertEquals(400, resp.status)
    }

    // ----- Models -----

    @Test
    fun `models returns single entry list with ctx modelId`() = runBlocking {
        val handlers = makeHandlers()
        val customCtx = HandlerContext(modelId = "/path/to/model.gguf", backendName = "llama")
        val resp = handlers.handleModels(customCtx) as? HandlerResponse.Json
            ?: error("expected Json response")
        assertEquals(200, resp.status)
        val obj = resp.body as JsonObject
        assertEquals("list", (obj["object"] as JsonPrimitive).content)
        val data = obj["data"] as JsonArray
        assertEquals(1, data.size)
        assertEquals("/path/to/model.gguf", ((data[0] as JsonObject)["id"] as JsonPrimitive).content)
    }
}
